zebra-state/src/sled_state.rs

```diff
@@ -178,11 +178,21 @@ impl FinalizedState {
         let height = queued_block.block.coinbase_height().unwrap();
         self.queued_by_prev_hash.insert(prev_hash, queued_block);
 
-        while let Some(queued_block) = self.queued_by_prev_hash.remove(&self.finalized_tip_hash()) {
+        while let Some((tip_height, tip_hash)) = self.tip() {
+            let queued_block =
+                if let Some(queued_block) = self.queued_by_prev_hash.remove(&tip_hash) {
+                    queued_block
+                } else {
+                    break;
+                };
+
             let height = queued_block
                 .block
                 .coinbase_height()
                 .expect("valid blocks must have a height");
+
+            assert_eq!(tip_height + 1, Some(height));
```

Comment:
There's a few checks missing here:
* the new block's parent hash is equal to the finalized tip hash
* if the state is empty, the new block's parent hash is equal to a null (all-zeroes) hash
* if the state is empty, the new block's height is equal to 0

The existing check is good though:
* the new block's height is one more than the finalized tip height

Also, these checks need to be in `commit_finalized_direct`, because that function is called by `StateService::queue_and_commit_non_finalized_blocks` without going through `queue_and_commit_finalized_blocks`.

And let's remove the corresponding TODO as part of this PR:
```
// TODO: check highest entry of hash_by_height as in RFC
```
It's in the wrong place - these checks should be outside the transaction.

Reply:
> the new block's parent hash is equal to the finalized tip hash

This check is unnecessary, blocks are only ever committed by dequeuing them based on their previous hash. It's impossible to ever commit blocks out of hash order by construction.

The checks aren't needed in `commit_finalized_direct`. The only other caller of `commit_finalized_direct` is the state service as part of `queue_and_commit_non_finalized_blocks` which independently guarantees that block hashes and heights are all valid, so the asserts would be redundant in that call graph.

Reply:
By definition, all of these assertions should be redundant, if the calling code is bug-free.

But the goal of this PR is to make sure that we never persist corrupted data to the sled database, *even if* we introduce bugs in the calling functions.

So these checks need to happen immediately before the data is written to the database, which is `commit_finalized_direct`. And we need to check all 4 cases.